File: ios-native/MSense/Services/PredictionService.swift
//
//  PredictionService.swift
//  MSense
//
//  Created by AI Assistant on 2025-08-08.
//

import Foundation
import Combine

class PredictionService: ObservableObject {
    @Published var predictions: [Double] = []
    @Published var nextMinutes: [Int] = [15, 30, 45, 60]
    @Published var historicalData: [FatigueDataPoint] = []
    @Published var isConnected: Bool = false
    @Published var isLoading: Bool = true
    
    private let baseURL = "http://localhost:5001" // Local prediction service
    private var pollingTimer: Timer?
    private let session = URLSession.shared
    
    init() {
        loadHistoricalData()
    }
    
    deinit {
        stopPolling()
    }
    
    // MARK: - Public Methods
    
    func startPolling() {
        isConnected = false
        getCurrentPredictions()
        
        // Poll every 15 minutes (900 seconds) to match server update interval
        pollingTimer = Timer.scheduledTimer(withTimeInterval: 900, repeats: true) { [weak self] _ in
            self?.getCurrentPredictions()
        }
    }
    
    func stopPolling() {
        pollingTimer?.invalidate()
        pollingTimer = nil
        isConnected = false
    }
    
    func getCurrentPredictions() {
        guard let url = URL(string: "\(baseURL)/current-predictions") else {
            print("Invalid URL")
            isConnected = false
            isLoading = false
            return
        }
        
        var request = URLRequest(url: url)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.timeoutInterval = 10
        
        session.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                self?.handlePredictionResponse(data: data, response: response, error: error)
            }
        }.resume()
    }
    
    func getHistoricalData(timeRange: TimeRange) -> [FatigueDataPoint] {
        if timeRange == .future {
            return historicalData
        }
        
        let cutoffTime = Date().addingTimeInterval(-timeRange.duration)
        return historicalData.filter { $0.timestamp > cutoffTime }
    }
    
    // MARK: - Private Methods
    
    private func handlePredictionResponse(data: Data?, response: URLResponse?, error: Error?) {
        isLoading = false
        
        if let error = error {
            print("Error fetching predictions: \(error)")
            isConnected = false
            return
        }
        
        guard let data = data else {
            print("No data received")
            isConnected = false
            return
        }
        
        do {
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            
            guard let status = json?["status"] as? String,
                  status == "success",
                  let predictionsArray = json?["predictions"] as? [Any] else {
                print("Invalid response format")
                isConnected = false
                return
            }
            
            // Convert predictions to doubles
            let newPredictions = predictionsArray.compactMap { prediction -> Double? in
                if let doubleValue = prediction as? Double {
                    return doubleValue
                } else if let intValue = prediction as? Int {
                    return Double(intValue)
                }
                return nil
            }
            
            let newNextMinutes = json?["next_minutes"] as? [Int] ?? [15, 30, 45, 60]
            
            self.predictions = newPredictions
            self.nextMinutes = newNextMinutes
            self.isConnected = true
            
            // Store current fatigue level in historical data (only if this is new data)
            if let currentFatigue = newPredictions.first {
                // Only add if this is different from the last data point (avoid duplicates on refresh)
                if historicalData.isEmpty || abs(historicalData.first!.value - currentFatigue) > 0.1 {
                    addHistoricalDataPoint(value: currentFatigue)
                }
            }
            
            // Check for notification triggers
            NotificationService.shared.checkForNotificationTriggers(
                predictions: newPredictions,
                nextMinutes: newNextMinutes
            )
            
            print("Received predictions: \(newPredictions) for minutes: \(newNextMinutes)")
            
        } catch {
            print("Error parsing JSON: \(error)")
            isConnected = false
        }
    }
    
    private func addHistoricalDataPoint(value: Double) {
        let dataPoint = FatigueDataPoint(timestamp: Date(), value: value)
        
        // Insert at beginning (newest first)
        historicalData.insert(dataPoint, at: 0)
        
        // Limit to 96 data points (24 hours at 15-minute intervals)
        if historicalData.count > 96 {
            historicalData = Array(historicalData.prefix(96))
        }
        
        saveHistoricalData()
        
        print("Added historical data point: \(value) at \(dataPoint.timestamp)")
    }
    
    private func loadHistoricalData() {
        guard let data = UserDefaults.standard.data(forKey: "historical_fatigue_data") else {
            print("No historical data found in UserDefaults")
            return
        }
        
        do {
            let decoder = JSONDecoder()
            historicalData = try decoder.decode([FatigueDataPoint].self, from: data)
            
            // Sort by timestamp (newest first)
            historicalData.sort { $0.timestamp > $1.timestamp }
            
            print("Loaded \(historicalData.count) historical data points")
            
        } catch {
            print("Error loading historical data: \(error)")
        }
    }
    
    private func saveHistoricalData() {
        do {
            let encoder = JSONEncoder()
            let data = try encoder.encode(historicalData)
            UserDefaults.standard.set(data, forKey: "historical_fatigue_data")
        } catch {
            print("Error saving historical data: \(error)")
        }
    }
}

// MARK: - Mock Implementation for Testing
extension PredictionService {
    func useMockData() {
        predictions = [5.0, 6.0, 7.0, 8.0]
        nextMinutes = [15, 30, 45, 60]
        isConnected = true
        isLoading = false
        
        // Generate mock historical data
        let now = Date()
        historicalData = (0..<24).map { index in
            FatigueDataPoint(
                timestamp: now.addingTimeInterval(TimeInterval(-index * 3600)), // Every hour for 24 hours
                value: 5.0 + Double(index % 5) // Values between 5-9
            )
        }
        
        saveHistoricalData()
    }
}

File: flutter-app/lib/services/prediction_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

class FatigueDataPoint {
  final DateTime timestamp;
  final double value;

  FatigueDataPoint({required this.timestamp, required this.value});

  Map<String, dynamic> toJson() => {
    'timestamp': timestamp.toIso8601String(),
    'value': value,
  };

  factory FatigueDataPoint.fromJson(Map<String, dynamic> json) {
    return FatigueDataPoint(
      timestamp: DateTime.parse(json['timestamp']),
      value: json['value'],
    );
  }
}

class PredictionService {
  // Singleton instance for global access
  static PredictionService instance = PredictionService();
  
  static const String baseUrl = 'http://10.0.2.2:5000';  // Special Android emulator localhost
  final _client = http.Client();
  Timer? _pollingTimer;
  bool _isPolling = false;
  
  // Historical data storage
  List<FatigueDataPoint> _historicalData = [];
  
  // Callbacks
  Function(List<double>, List<int>)? onPredictionsUpdated;
  Function(bool)? onConnectionStatusChanged;
  Function(List<FatigueDataPoint>)? onHistoricalDataUpdated;

  PredictionService() {
    _loadHistoricalData();
  }

  Future<void> _loadHistoricalData() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final jsonData = prefs.getString('historical_fatigue_data');
      
      if (jsonData != null) {
        final List<dynamic> decoded = json.decode(jsonData);
        _historicalData = decoded
            .map((item) => FatigueDataPoint.fromJson(item))
            .toList();
        
        // Sort by timestamp (newest first)
        _historicalData.sort((a, b) => b.timestamp.compareTo(a.timestamp));
        
        // Print all historical data for debugging
        print('=== LOADED HISTORICAL DATA ===');
        print('Total data points: ${_historicalData.length}');
        for (var i = 0; i < _historicalData.length; i++) {
          print('[$i] ${_historicalData[i].timestamp} - Value: ${_historicalData[i].value}');
        }
        print('==============================');
        
        // Notify listeners
        onHistoricalDataUpdated?.call(_historicalData);
      } else {
        print('=== NO HISTORICAL DATA FOUND IN SHARED PREFERENCES ===');
      }
    } catch (e) {
      print('Error loading historical data: $e');
    }
  }

  Future<void> _saveHistoricalData() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final jsonData = json.encode(_historicalData.map((dp) => dp.toJson()).toList());
      await prefs.setString('historical_fatigue_data', jsonData);
    } catch (e) {
      print('Error saving historical data: $e');
    }
  }

  List<FatigueDataPoint> getHistoricalData({Duration? timeRange}) {
    if (timeRange == null) {
      return List.from(_historicalData);
    }
    
    final cutoffTime = DateTime.now().subtract(timeRange);
    return _historicalData
        .where((dataPoint) => dataPoint.timestamp.isAfter(cutoffTime))
        .toList();
  }

  Future<void> startPolling() async {
    if (_isPolling) return;
    _isPolling = true;
    onConnectionStatusChanged?.call(false);  // Start with disconnected state
    
    // Get initial predictions
    await getCurrentPredictions();
    
    // Start polling every 15 minutes (900 seconds) to match the server update interval
    _pollingTimer = Timer.periodic(const Duration(seconds: 900), (_) {
      getCurrentPredictions();
    });
  }

  Future<void> getCurrentPredictions() async {
    try {
      print('Fetching predictions from $baseUrl/current-predictions');
      final response = await _client.get(
        Uri.parse('$baseUrl/current-predictions'),
        headers: {
          'Accept': 'application/json',
          'Access-Control-Allow-Origin': '*',
        },
      ).timeout(const Duration(seconds: 10)); // Add timeout
      
      print('Response status: ${response.statusCode}');
      print('Response body: ${response.body}');
      
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        
        if (data['status'] == 'success' && data['predictions'] != null) {
          // Convert integers to doubles if needed
          final predictions = data['predictions'].map<double>((pred) => 
            pred is int ? pred.toDouble() : (pred is double ? pred : 0.0)
          ).toList();
          
          final nextMinutes = List<int>.from(data['next_minutes'] ?? [15, 30, 45, 60]);
          print('Received predictions: $predictions for minutes: $nextMinutes');
          
          // Store the current fatigue level in historical data
          if (predictions.isNotEmpty) {
            final currentFatigue = predictions[0];
            _addHistoricalDataPoint(currentFatigue);
          }
          
          onPredictionsUpdated?.call(predictions, nextMinutes);
          onConnectionStatusChanged?.call(true);
        } else {
          print('Invalid response format: $data');
          onConnectionStatusChanged?.call(false);
        }
      } else {
        print('Error response: ${response.statusCode}');
        onConnectionStatusChanged?.call(false);
      }
    } catch (e) {
      print('Error fetching predictions: $e');
      onConnectionStatusChanged?.call(false);
    }
  }

  void _addHistoricalDataPoint(double value) {
    final now = DateTime.now();
    final dataPoint = FatigueDataPoint(timestamp: now, value: value);
    
    print('=== ADDING NEW HISTORICAL DATA POINT ===');
    print('Timestamp: $now - Value: $value');
    
    // Add to the beginning of the list (newest first)
    _historicalData.insert(0, dataPoint);
    
    // Limit the size of historical data (keep last 24 hours = 96 data points at 15-minute intervals)
    if (_historicalData.length > 96) {
      _historicalData = _historicalData.sublist(0, 96);
    }
    
    print('Total historical data points after adding: ${_historicalData.length}');
    
    // Save to SharedPreferences
    _saveHistoricalData();
    
    // Notify listeners
    onHistoricalDataUpdated?.call(_historicalData);
  }

  void stopPolling() {
    _pollingTimer?.cancel();
    _pollingTimer = null;
    _isPolling = false;
    onConnectionStatusChanged?.call(false);
  }
}
